arch/arm/src/armv7-m: Add support of nested interrupt in flat binary

drashti304 opened the pull request:
This patch adds support of nested interrupt in flat binary
as well as move logic of incrementing g_nestlevel and storing
current_regs value in critical area under exception_common function

Signed-off-by: Drashti Parikh <p.drashti@partner.samsung.com>

seinfra commented:
Target : [4bc47221349c12d8528354fec68d8dbcbde6c8bf] - Code Rule Check (C++) OK.

seinfra commented:
Target : [4bc47221349c12d8528354fec68d8dbcbde6c8bf] - Code Rule Check OK.

seinfra commented:
Target : [3e983f38a79a74a43418eb2e761f0bf12b4bf757] - Code Rule Check OK.

seinfra commented:
Target : [3e983f38a79a74a43418eb2e761f0bf12b4bf757] - Code Rule Check (C++) OK.

sunghan-chang commented on os/arch/arm/src/armv7-m/up_doirq.c:
```diff
@@ -117,13 +117,17 @@ uint32_t *up_doirq(int irq, uint32_t *regs)
 	}
 
-	g_nestlevel++;
+	g_nestlevel++;*/
```
Do we need to keep them with commented?

drashti304 replied:
Not sure, I though to keep this code and added the comment stating "To prevent higher priority interrupt corrupt current_regs, Moved below code in exception_common under critical area"

But I am seeking your suggestion on that, what is better to do? remove code or add comment ?

sunghan-chang replied:
I think you can make a separated commit only for this. It is easier to know why we modify this.

sunghan-chang replied:
I mean we could remove or move with separated commit. not for commenting.

drashti304 commented:
> Do you include the fix of [Samsung/TizenRT-Mediatek#380 (comment)](https://github.com/Samsung/TizenRT-Mediatek/issues/380#issuecomment-539499486)?

Yes @sunghan-chang , I have addressed the issues mentioned in Samsung/TizenRT-Mediatek#380 (comment)

sunghan-chang commented:
Doesn't it necessary for ARMv7-R?

seinfra commented:
Target : [462797af6689ff6950abd11c14c799a578f8ee50] - Code Rule Check OK.

sunghan-chang commented on os/arch/arm/src/armv7-m/up_lazyexception.S:
```diff
@@ -245,11 +246,12 @@ exception_common:
 	 * to the interrupt stack; if g_nestlevel is greater than zero, then do
 	 * not switch stacks. In this latter case, we are already using the interrupt stack */
-
+#if CONFIG_ARCH_NESTED_INTERRUPT
```
`#ifdef`

drashti304 replied:
Done thanks

drashti304 commented:
> Isn't it necessary for ARMv7-R?

As we are planning to go with armv7-m chip sets, I have add support for armv7-m architecture only.
for armv7-r many things needs to be changed so if in future needs arise we will add it.

seinfra commented:
Target : [94497e06dec33680f1414c089a34a3e823f8c155] - Code Rule Check OK.

seinfra commented:
Target : [94497e06dec33680f1414c089a34a3e823f8c155] - Code Rule Check (C++) OK.
